File: Homework/2022_09_16/math_stuff.py
#!/usr/bin/env python3
# coding: utf8

# Homework 2:
#   Math stuff yaaay!!
# In this excersize we have to find 3 formulas, and implement them in python

from functools import reduce
from math import sqrt
from typing import Dict, List, Tuple, AnyStr


def masodfoku(a, b, c) -> Tuple[float, float]: # Másodfokú egyenlet, doesn't handle edge cases!!
    return (-b+sqrt(b**2-4*a*c)/2*a), (-b-sqrt(b**2-4*a*c))/2*a

# Weighted average formula
# Input is a list of dictionary that has a weight and a value
def weightedaverage(to_be_averaged: List[Dict[AnyStr, int]]):
    weighted_value = 0
    for x in to_be_averaged:
        weighted_value += x["weight"]*x["value"]
    
    sum_of_weight = 0
    for x in to_be_averaged:
        sum_of_weight += x["weight"]

    return weighted_value / sum_of_weight

# Standard deviation calculation
# √(Sum(|x-averge(x)|^2)/n)
def standard_deviation(l: List[int]):
    average = sum(l) / len(l)
    s = 0.0
    for x in l:
        s += abs(x-average)**2
    return sqrt(s / len(l))

def main():
    print(masodfoku(2, 10, 3))
    print(weightedaverage([{"weight": 25, "value": 75}, {"weight": 30, "value": 60}]))
    print(standard_deviation([9, 2, 5, 4, 12, 7, 8, 11, 9, 3, 7, 4, 12, 5, 4, 10, 9, 6, 9, 4]))


if __name__ == "__main__":
    main()

File: Homework/2022_09_16/rust_homework/src/bin/math_stuff.rs
fn masodfoku(a: f64, b: f64, c: f64) -> (f64, f64) {
    return (
        -b + (b.powi(2) - 4.0 * a * c).sqrt(),
        -b - (b.powi(2) - 4.0 * a * c).sqrt(),
    );
}

struct WeightAve {
    weight: i64,
    value: i64,
}

fn weighted_average(l: &Vec<WeightAve>) -> f64 {
    let weighted_value = l.iter().fold(0, |acc: i64, x| acc + x.weight * x.value);
    let weight_sum = l.iter().fold(0, |acc, x| acc + x.weight);

    return weighted_value as f64 / weight_sum as f64;
}

fn standard_deviation(l: &Vec<i32>) -> f64 {
    let average = l.iter().fold(0, |acc, x| acc + x) as f64 / l.len() as f64;
    let mut s = 0.0;
    for x in l {
        s += ((*x as f64 - average) as f64).powi(2)
    }
    (s / l.len() as f64).sqrt()
}

fn main() {
    println!("{:?}", masodfoku(2.0, 10.0, 3.0));
    println!(
        "{}",
        weighted_average(&vec![
            WeightAve {
                weight: 25,
                value: 75
            },
            WeightAve {
                weight: 30,
                value: 60
            }
        ])
    );
    println!(
        "{}",
        standard_deviation(&vec![
            9, 2, 5, 4, 12, 7, 8, 11, 9, 3, 7, 4, 12, 5, 4, 10, 9, 6, 9, 4
        ])
    )
}
